astor/code_gen.py

```diff
@@ -50,22 +50,30 @@ def to_source(node, indent_with=' ' * 4, add_line_information=False,
                                 pretty_string)
     generator.visit(node)
     generator.result.append('\n')
-    return pretty_source(str(s) for s in generator.result)
+    if set(generator.result[0]) == '\n':
+        generator.result[0] = ''
+    return pretty_source(generator.result)
 
 
-def set_precedence(value, *nodes):
-    """Set the precedence (of the parent) into the children.
-    """
-    if isinstance(value, ast.AST):
-        value = get_op_precedence(value)
-    for node in nodes:
-        if isinstance(node, ast.AST):
-            node._pp = value
-        elif isinstance(node, list):
-            set_precedence(value, *node)
-        else:
-            assert node is None, node
+def precedence_setter(AST=ast.AST, get_op_precedence=get_op_precedence,
+                      isinstance=isinstance, list=list):
 
+    def set_precedence(value, *nodes):
+        """Set the precedence (of the parent) into the children.
+        """
+        if isinstance(value, AST):
+            value = get_op_precedence(value)
+        for node in nodes:
+            if isinstance(node, AST):
+                node._pp = value
+            elif isinstance(node, list):
+                set_precedence(value, *node)
+            else:
+                assert node is None, node
+
+    return set_precedence
+
+set_precedence = precedence_setter()
```

Comment:
set_precedence is called a _lot_, so this minimal closure to reduce attribute lookups is (IMO) worth it.  It's pretty standalone.